xarray/core/dataarray.py

```diff
@@ -1736,5 +1736,93 @@ def dot(self, other):
 
         return type(self)(new_data, new_coords, new_dims)
 
+    def quantile(self, q, dim=None, axis=None, interpolation='linear'):
```

Comment:
It would be nice to also have this method for `Dataset` (or maybe even for `Variable`, too). Take a look at `Dataset.reduce` for an example.

To avoid duplicate implementation, many `DataArray` methods have the core of their logic implemented for Dataset and then use `_to_temp_dataset()` and `_to_from_dataset()` to convert back and forth.